repository/copier.go

```diff
@@ -113,126 +197,154 @@ func (c *HDFSCopier) freeClient(err error) (wrapped error) {
 	return
 }
 
-func (c *HDFSCopier) deleteIfExists(file string) error {
-	if _, err := c.client.Stat(file); os.IsNotExist(err) {
-		return nil
-	} else if err != nil {
-		return err
-	}
-
-	return c.client.Remove(file)
-}
-
-// CopyToRemote copies from the provided billy Filesystem to HDFS. If the file exists on HDFS it will be overridden.
-// If other writer is actually copying the same file to HDFS this method will throw an error because the WORM principle
-// (Write Once Read Many).
-func (c *HDFSCopier) CopyToRemote(src, dst string, localFs billy.Filesystem) (err error) {
+func (fs *hdfsFs) WriteTo(path string) (w io.WriteCloser, err error) {
 	defer func() {
-		err = c.freeClient(err)
+		err = fs.freeClient(err)
 	}()
 
-	bDst := addBucketName(dst, c.bucketSize)
-	p := path.Join(c.base, bDst)
-	if err := c.initializeClient(); err != nil {
-		return err
+	if err = fs.initializeClient(); err != nil {
+		return
 	}
 
-	pCopy := path.Join(c.tmpBase, dst+".copy")
+	if err = fs.client.MkdirAll(filepath.Dir(path), os.FileMode(0644)); err != nil {
+		return
+	}
 
-	lf, err := localFs.Open(src)
-	if os.IsNotExist(err) {
-		return nil
+	if err = fs.DeleteIfExists(path); err != nil {
+		return
 	}
+
+	w, err = fs.client.Create(path)
 	if err != nil {
-		return err
+		return
 	}
-	defer checkClose(lf, &err)
 
-	if err := c.client.MkdirAll(path.Dir(p), os.FileMode(0644)); err != nil {
-		return err
-	}
+	return
+}
 
-	if err := c.client.MkdirAll(path.Dir(pCopy), os.FileMode(0644)); err != nil {
-		return err
-	}
+func (fs *hdfsFs) Rename(src, dst string) (err error) {
+	defer func() {
+		err = fs.freeClient(err)
+	}()
 
-	// TODO to avoid this, we should implement a 'truncate' flag in 'client.Create' method
-	err = c.deleteIfExists(pCopy)
-	if err != nil {
+	if err = fs.client.MkdirAll(filepath.Dir(dst), os.FileMode(0644)); err != nil {
 		return err
 	}
 
-	rf, err := c.client.Create(pCopy)
-	if err != nil {
-		return err
-	}
+	err = fs.client.Rename(src, dst)
+	return
+}
 
-	// Delete temporary file in case the process is stopped while copying
+func (fs *hdfsFs) DeleteIfExists(file string) (err error) {
 	defer func() {
-		rf.Close()
-		c.deleteIfExists(pCopy)
+		err = fs.freeClient(err)
 	}()
 
-	_, err = io.Copy(rf, lf)
-
-	checkClose(rf, &err)
-
-	if err != nil {
-		c.client.Remove(pCopy)
+	if _, err = fs.client.Stat(file); os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
 		return err
 	}
 
-	err = c.client.Rename(pCopy, p)
-
-	if err != nil {
+	if err = fs.client.Remove(file); err != nil {
 		return err
 	}
 
 	return
 }
 
-func (c *HDFSCopier) initializeClient() (err error) {
-	if c.client != nil {
-		return
-	}
-	c.client, err = hdfs.New(c.url)
+func (fs *hdfsFs) TempDir() string {
+	return fs.tmpBase
+}
 
-	return
+func (fs *hdfsFs) Base() string {
+	return fs.base
 }
 
 func checkClose(c io.Closer, err *error) {
 	if cerr := c.Close(); cerr != nil && *err == nil {
 		*err = cerr
 	}
 }
+
+const copySize = 32 * 1024
```

Comment:
I'm just curious - why this number? As far as I know maximum throughput we can get with 64K buffer.
On ther other hand, long time ago on linux I was used to use buffer size 4K because page size is/was 4K.

Reply:
It's the same number `io.Copy` uses. But if you guys think we should change the number let's agree on a number. 64K sounds good.